internal/bft/viewchanger.go

```
@@ -448,42 +448,149 @@ func (v *ViewChanger) getInFlight(lastDecision *protos.Proposal) *protos.Proposa
+
+	if lastDecisionMD.LatestSequence > mySequence+1 { // this is a decision in the future, ignoring since the node might not have the right configuration to validate
+		v.Logger.Debugf("Node %d got %s from %d, but the last decision seq %d is greater than this node's current sequence %d", v.SelfID, signedViewDataToString(svd), sender, lastDecisionMD.LatestSequence, mySequence)
+		return false, 0
+	}
+	if lastDecisionMD.LatestSequence < mySequence { // this is a decision in the past, ignoring since the node might not have the right configuration to validate
+		v.Logger.Debugf("Node %d got %s from %d, but the last decision seq %d is lower than this node's current sequence %d", v.SelfID, signedViewDataToString(svd), sender, lastDecisionMD.LatestSequence, mySequence)
+		return false, 0
+	}
+
+	if lastDecisionMD.LatestSequence == mySequence { // just make sure that we have the same last decision, can't verify the signatures of this last decision since this might have been a reconfiguration
+		// the signature on this message can be verified
+		if svd.Signer != sender {
+			v.Logger.Warnf("Node %d got %s from %d, but signer %d is not the sender %d", v.SelfID, signedViewDataToString(svd), sender, svd.Signer, sender)
+			return false, 0
+		}
+		if err := v.Verifier.VerifySignature(types.Signature{ID: svd.Signer, Value: svd.Signature, Msg: svd.RawViewData}); err != nil {
+			v.Logger.Warnf("Node %d got %s from %d, but signature is invalid, error: %v", v.SelfID, signedViewDataToString(svd), sender, err)
+			return false, 0
+		}
+
+		// compare the last decision itself
+		if !proto.Equal(vd.LastDecision, myLastDecision) {
+			v.Logger.Warnf("Node %d got %s from %d, they are at the same sequence but the last decisions are not equal", v.SelfID, signedViewDataToString(svd), sender)
+			return false, 0
+		}
+
+		return true, lastDecisionMD.LatestSequence
+	}
+
+	if lastDecisionMD.LatestSequence == mySequence+1 {
```

perhaps check the complement condition and fail fast (return false) instead of indenting this whole block